mithril-aggregator/src/database/provider/cardano_transaction.rs

```diff
@@ -141,41 +160,39 @@ impl<'client> InsertCardanoTransactionProvider<'client> {
     }
 
     fn get_insert_condition(&self, record: &CardanoTransactionRecord) -> StdResult<WhereCondition> {
-        let expression =
-            "(transaction_hash, block_number, immutable_file_number) values (?1, ?2, ?3)";
-        let parameters = vec![
-            Value::String(record.transaction_hash.clone()),
-            Value::Integer(record.block_number.try_into()?),
-            Value::Integer(record.immutable_file_number.try_into()?),
-        ];
-
-        Ok(WhereCondition::new(expression, parameters))
+        self.get_insert_many_condition(vec![record.clone()])
     }
 
     fn get_insert_many_condition(
         &self,
         transactions_records: Vec<CardanoTransactionRecord>,
-    ) -> WhereCondition {
-        let columns = "(transaction_hash, block_number, immutable_file_number)";
-        let values_columns: Vec<&str> = repeat("(?*, ?*, ?*)")
+    ) -> StdResult<WhereCondition> {
+        fn map_record(record: CardanoTransactionRecord) -> StdResult<Vec<Value>> {
+            Ok(vec![
+                Value::String(record.transaction_hash),
+                Value::Integer(record.block_number.try_into().unwrap()),
+                Value::Integer(record.slot_number.try_into()?),
+                Value::String(record.block_hash.clone()),
+                Value::Integer(record.immutable_file_number.try_into().unwrap()),
+            ])
+        }
+
+        let columns =
+            "(transaction_hash, block_number, slot_number, block_hash, immutable_file_number)";
+        let values_columns: Vec<&str> = repeat("(?*, ?*, ?*, ?*, ?*)")
             .take(transactions_records.len())
             .collect();
 
-        let values: Vec<Value> = transactions_records
-            .into_iter()
-            .flat_map(|record| {
-                vec![
-                    Value::String(record.transaction_hash),
-                    Value::Integer(record.block_number.try_into().unwrap()),
-                    Value::Integer(record.immutable_file_number.try_into().unwrap()),
-                ]
-            })
-            .collect();
+        // TODO see if we can find another way to do it
+        let values: StdResult<Vec<Vec<Value>>> =
+            transactions_records.into_iter().map(map_record).collect();
 
-        WhereCondition::new(
+        let values: Vec<Value> = values?.into_iter().flatten().collect();
```

Comment:
This code is less readable than previous one and requires to create 2 vec instead of one.
Why didn't you use `flat_map` here?

Reply:
We can't use a `flat_map` as that it will flatten the result hence the comment, this is the only way that we find in order to have a Result over the Values conversion (or it would still be using unwraps).

Reply:
Have you tried a `try_fold`?